### DISCUSSION TITLE
Difference between LOCAL, BROKER_ID and BROKER_USERNAME on username template mapper

### DISCUSSION THREAD

**Discussion by oculos** [keycloak/keycloak#29849] [Q&A]
Hi,

I can't find any good documentation on what's the difference between `LOCAL`, `BROKER_ID` and `BROKER_USERNAME` on the documentation.

The only explanation I found is on the tooltip:

> Destination field for the mapper. LOCAL (default) means that the changes are applied to the username stored in local database upon user import. BROKER_ID and BROKER_USERNAME means that the changes are stored into the ID or username used for federation user lookup, respectively.

But there are things it doesn't say:

- is there a priority among these? I noticed, for example, that if I have the three of them configured, changing `BROKER_USESRNAME` to map to another claim won't update the Broker username, even with `force` on the sync option.
- Can they be changed arbitrarily? I noticed that, changing Broker_ID to another attribute will break authentication for existing accounts. I guess that because Broker_ID is stored on the database and connected to a user, changing it leads to Keycloak trying to link an already linked account. However, changing "LOCAL" seems to reflect the changes when using force.

It would be very nice to have an idea of when to use these mappers and the consequences of choosing them.

**Comment by omasseau** [comment]
Agree, honestly I don't understand anything from the description of the popup, and in which case we should use a value or another.
